feat: Offline Support with nested schema

Lavanyagaur22 opened the pull request:
### Description
Inside the `interceptAsync()` of `ConflictInterceptor` class, I have checked for the network connection. If the network connection is not there, and if the request operation is a mutation, I store them in an array list as of now. 
Whenever the net comes, we wait for a mutation request from the user side. Upon getting it, we fetch the mutation requests stored in the list and replicate them back to the server. 
We are not providing UI bindings to the user currently.

### How to test it?
- Start server.
- Go offline. 
- Make some mutations. (For updating a task, click on the task and provide the details in it. )
- Come online.
- Make any mutation. And you will see the results of the mutations done when offline along with the result of the mutation done when the network connection is regained.

**Also trying to study request so that we can parse it accordingly and store important things in database and get rid of array list**

Lavanyagaur22 committed:
feat: Offline Support with nested schema

wtrocki commented on offix/src/main/java/org/aerogear/offix/OfflineCallback.kt:
```diff
@@ -0,0 +1,42 @@
+    override fun onResponse(response: ApolloInterceptor.InterceptorResponse) {
+        Log.d(TAG, "onResponse()")
+//
+//        /* Check if the conflict is present in the response of not using the ConflictResolutionHandler class.
```
??

Lavanyagaur22 committed:
added logs for debDugging

Lavanyagaur22 committed:
minor fix in interceptor

Lavanyagaur22 committed:
minor fix in flow of interceptor

Lavanyagaur22 committed:
calls working properly, minor nit needed in UI data

Lavanyagaur22 committed:
Offline support foreground implemented

wtrocki commented:
> We are not providing UI bindings to the user currently.

That is absolutely fine. How to test this PR?

Lavanyagaur22 commented:
@wtrocki 

> > We are not providing UI bindings to the user currently.
> 
> That is absolutely fine. How to test this PR?

### How to test it?
- Start server.
- Go offline.
- Make some mutations. (For updating a task, click on the task and provide the details in it. )
- Come online.
- Make any mutation. And you will see the results of the mutations done when offline along with the result of the mutation done when the network connection is regained.

harshithdwivedi commented:
@Lavanyagaur22 is this done? 
If so, we can drop the WIP tag maybe :)

harshithdwivedi reviewed (approved):
LGTM, good work.
P.S. Maybe go a little easy with the logs the next time hehe

wtrocki commented:
Went offline. Created couple items, edited them. Went online - Do see updates failed on server.

Second test. Created item online. Went offline done two edits. Back online.  Got conflict. 

Missing anything?

wtrocki commented:
I think I have mixed things up. Does offline create work for you?

Lavanyagaur22 committed:
fix: version added in createFragment

Lavanyagaur22 commented:
## 1. When offline:
- Updated 2 tasks: (Task 1 and Task 2)
- Then came online, and updated Task 3

### Result of the 3 mutation requests: 
<img width="1440" alt="Screenshot 2019-08-16 at 12 53 16 PM" src="https://user-images.githubusercontent.com/33238323/63151079-45920b00-c026-11e9-8d3a-e8bc312a38f3.png">

## 2. When offline:
@wtrocki Create Task is working fine.

Lavanyagaur22 commented:
> I think I have mixed things up. Does offline create work for you?

Yes, it's working. Maybe you can try it now once again, you might be ending up in null ptr exception because of the null in version field which might be resulting in the crashing of your app. Fixed it now.

Lavanyagaur22 commented:
> I think I have mixed things up. Does offline create work for you?

@wtrocki Yes, it's working. Maybe you can try it now once again, you might be ending up in null ptr exception because of the null in version field which might be resulting in the crashing of your app. Fixed it.

Lavanyagaur22 commented:
Also, @harshithdwivedi @wtrocki Our offline requests are working perfect and the results can be seen on the server too. But when the offline create mutations hit the server successfully, changes are not being reflected back in Apollo cache.

wtrocki commented:
Ok. So lack of item in cache can explain it. Do you think we can have subscriptions implemented to get updates to cache for now?

Lavanyagaur22 commented:
> Ok. So lack of item in cache can explain it. Do you think we can have subscriptions implemented to get updates to cache for now?

Yeah, I was about to open issue for using subscriptions in app like we were using earlier.

wtrocki commented:
Let's address Codacy issues and merge.

Lavanyagaur22 committed:
fix: apollo cache updating perfectly and fix codacy error

Lavanyagaur22 committed:
fix:codacy error
